pkg/controller/node/nodecontroller.go

```diff
@@ -544,6 +548,63 @@ func (nc *NodeController) monitorNodeStatus() error {
 	return nil
 }
 
+// monitorNodeTaints for all nodes compares taints with pods tolerations
+// and sends pods for eviction
+func (nc *NodeController) monitorNodeTaints() error {
+	nodes, err := nc.kubeClient.Core().Nodes().List(api.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodes.Items {
+		// node should be visited by monitorNodeStatus first, so that zonePodeEvictor will be initialized
+		if _, ok := nc.knownNodeSet[node.Name]; !ok {
+			continue
+		}
+		taints, err := api.GetTaintsFromNodeAnnotations(node.Annotations)
+		if len(taints) == 0 {
+			continue
+		}
+		pods, err := getPodsForANode(nc.kubeClient, node.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, pod := range pods.Items {
+			tolerations, err := api.GetTolerationsFromPodAnnotations(pod.Annotations)
+			if err != nil {
+				return err
+			}
+
+			if !tolerationsToleratesTaints(tolerations, taints) {
+				nc.evictPods(&node, pod)
+			} else {
+				nc.cancelPodsEviction(&node, pod)
+			}
+		}
+	}
+	return nil
+}
+
+func (nc *NodeController) evictPods(node *api.Node, pods ...api.Pod) {
+	for _, pod := range pods {
+		glog.Infof("pod %v:%v will be sent for eviction", pod.Namespace, pod.Name)
+		message := evictionMessage{pod.Name, pod.Namespace, node.UID}
+		nc.zonePodEvictor[utilnode.GetZoneKey(node)].Add(pod.Namespace+":"+pod.Name, message)
+	}
+}
+
+func (nc *NodeController) cancelPodsEviction(node *api.Node, pods ...api.Pod) {
+	for _, pod := range pods {
```

Comment:
Can this race with eviction in a way that causes the node controller to "forget" a pod that has been deleted already and needs to be drawn to conclusion?